File: examples/rust/player/agg/src/handlers/withdraw.rs
//! WithdrawFunds command handler.

use angzarr_client::proto::examples::{Currency, FundsWithdrawn, WithdrawFunds};
use angzarr_client::proto::{CommandBook, EventBook};
use angzarr_client::{new_event_book, pack_event, CommandRejectedError, CommandResult, UnpackAny};
use prost_types::Any;

use crate::state::PlayerState;

fn guard(state: &PlayerState) -> CommandResult<()> {
    if !state.exists() {
        return Err(CommandRejectedError::new("Player does not exist"));
    }
    Ok(())
}

fn validate(cmd: &WithdrawFunds, state: &PlayerState) -> CommandResult<i64> {
    let amount = cmd.amount.as_ref().map(|c| c.amount).unwrap_or(0);
    if amount <= 0 {
        return Err(CommandRejectedError::new("amount must be positive"));
    }
    if amount > state.available_balance() {
        return Err(CommandRejectedError::new("insufficient available balance"));
    }
    Ok(amount)
}

fn compute(cmd: &WithdrawFunds, state: &PlayerState, amount: i64) -> FundsWithdrawn {
    let new_balance = state.bankroll - amount;
    FundsWithdrawn {
        amount: cmd.amount.clone(),
        new_balance: Some(Currency {
            amount: new_balance,
            currency_code: "CHIPS".to_string(),
        }),
        withdrawn_at: Some(angzarr_client::now()),
    }
}

pub fn handle_withdraw_funds(
    command_book: &CommandBook,
    command_any: &Any,
    state: &PlayerState,
    seq: u32,
) -> CommandResult<EventBook> {
    let cmd: WithdrawFunds = command_any
        .unpack()
        .map_err(|e| CommandRejectedError::new(format!("Failed to decode command: {}", e)))?;

    guard(state)?;
    let amount = validate(&cmd, state)?;

    let event = compute(&cmd, state, amount);
    let event_any = pack_event(&event, "examples.FundsWithdrawn");

    Ok(new_event_book(command_book, seq, event_any))
}

File: examples/go/player/agg/handlers/withdraw.go
package handlers

import (
	"time"

	angzarr "github.com/benjaminabbitt/angzarr/client/go"
	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"github.com/benjaminabbitt/angzarr/client/go/proto/examples"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

func guardWithdrawFunds(state PlayerState) error {
	if !state.Exists() {
		return angzarr.NewCommandRejectedError("Player does not exist")
	}
	return nil
}

func validateWithdrawFunds(cmd *examples.WithdrawFunds, state PlayerState) (int64, error) {
	amount := int64(0)
	if cmd.Amount != nil {
		amount = cmd.Amount.Amount
	}
	if amount <= 0 {
		return 0, angzarr.NewCommandRejectedError("amount must be positive")
	}
	if amount > state.AvailableBalance() {
		return 0, angzarr.NewCommandRejectedError("insufficient available balance")
	}
	return amount, nil
}

func computeFundsWithdrawn(cmd *examples.WithdrawFunds, state PlayerState, amount int64) *examples.FundsWithdrawn {
	newBalance := state.Bankroll - amount
	return &examples.FundsWithdrawn{
		Amount:      cmd.Amount,
		NewBalance:  &examples.Currency{Amount: newBalance, CurrencyCode: "CHIPS"},
		WithdrawnAt: timestamppb.New(time.Now()),
	}
}

// HandleWithdrawFunds handles the WithdrawFunds command.
func HandleWithdrawFunds(
	commandBook *pb.CommandBook,
	commandAny *anypb.Any,
	state PlayerState,
	seq uint32,
) (*pb.EventBook, error) {
	var cmd examples.WithdrawFunds
	if err := proto.Unmarshal(commandAny.Value, &cmd); err != nil {
		return nil, err
	}

	if err := guardWithdrawFunds(state); err != nil {
		return nil, err
	}
	amount, err := validateWithdrawFunds(&cmd, state)
	if err != nil {
		return nil, err
	}

	event := computeFundsWithdrawn(&cmd, state, amount)

	eventAny, err := anypb.New(event)
	if err != nil {
		return nil, err
	}

	return angzarr.NewEventBook(commandBook.Cover, seq, eventAny), nil
}
